### PR TITLE
Support BTreeMap and fix rpc optional

### PR DESCRIPTION
**Pull request by Neopallium**
Closes issue #1723

### CODE REVIEW

**Comment by TarikGul** [comment]
@Neopallium I have this on my list to review, and will do it as soon as possible so we can get this in.

File: packages/react-params/src/Param/BTreeMap.tsx [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+  useEffect((): void => {
+    const output = new Map();
+    let isValid = true;
+
+    for (const entry of values) {
+      const [key, value] = entry.value as RawParam[];
+
+      if (output.has(key)) {
+        isValid = false;
+        console.error('BTreeMap: Duplicate key', key);
```
Comment (TarikGul) [inline packages/react-params/src/Param/BTreeMap.tsx:104]:
```suggestion
        console.error('BTreeMap: Duplicate key ', key);
```

File: packages/react-params/src/Param/Vector.tsx [hunk 2]
```diff
@@ -61,9 +60,9 @@ function Vector ({ className = '', defaultValue, isDisabled = false, label, onCh
   useEffect((): void => {
     inputParams.length &&
       setParams((prev) =>
-        getParams(inputParams, prev, isDisabled ? (defaultValue.value as RawParam[] || []).length : count)
+        getParams(inputParams, prev, isDisabled ? values.length : count)
```
Comment (TarikGul) [inline packages/react-params/src/Param/Vector.tsx:63]: Ahh nice catch here 👍!

File: packages/page-rpc/src/Rpc/Selection.tsx [hunk 3]
```diff
@@ -72,7 +72,7 @@ function Selection ({ queueRpc }: Props): React.ReactElement<Props> {
     (): void => queueRpc({
       rpc,
       values: values
-        .filter(({ value }) => !isNull(value))
+        .filter(({ value }, idx) => !rpc.params[idx].isOptional || !isNull(value))
```
Comment (TarikGul) [inline packages/page-rpc/src/Rpc/Selection.tsx:75]: Apologies but can you give a brief explanation as to what this is fixing? Thanks again.

**Review by TarikGul — CHANGES_REQUESTED** [review]
Overall, looks super solid Thanks so much for the PR! The BTreeMap Component is spot on, and the vector fix is awesome.

Just a small nit with the console, and a quick question about the rpc optional fix. Other than that it should be ready to go.

File: packages/page-rpc/src/Rpc/Selection.tsx (hunk 3 above)
Comment (Neopallium) [reply packages/page-rpc/src/Rpc/Selection.tsx:75]: The `isOptional` setting only really works for last parameters, not for a mix of Option and non-Option parameters.  So the fix keeps null values for `isOptional=false` parameters.

We are using this fix here: https://mainnet-app.polymesh.network/#/rpc
Open the `asset.canTransferGranular` endpoint and just click the submit button (don't need to fill in values).

To see the error try the same endpoint using the Polkadot.js.org app without the fix: https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fmainnet-rpc.polymesh.network#/rpc

The error is:
```
1: asset.canTransferGranular: CanTransferGranularReturn
Expected 7 parameters (1 optional), 4 found instead
```

**Comment by TarikGul** [comment]
Just needs a `yarn lint`

File: packages/page-rpc/src/Rpc/Selection.tsx (hunk 3 above)
Comment (TarikGul) [reply packages/page-rpc/src/Rpc/Selection.tsx:75]: Ahh okay, thanks so much for the explanation. Makes sense

**Comment by Neopallium** [comment]
`lint` applied.

**Review by TarikGul — APPROVED** [review]

**Pull request merged**